Add OTE SDK training test helpers

LeonidBeynenson opened the pull request.

LeonidBeynenson committed:
Add algo backend test helpers for reallife tests.

Switch mmdetection submodule to the current state of the branch
lb/moving_reallive_training_tests.
The branch is intended to be a reference code of using algo backend test
helpers for reallife tests.

LeonidBeynenson committed:
Move common fixtures and pytest code insertions to OTE SDK test helpers

LeonidBeynenson committed:
Finishing moving OTE test helpers to OTE SDK

LeonidBeynenson committed:
Update docstrings in OTE SDK test helpers

LeonidBeynenson commented:
Note that this PR depends on the mmdetection PR
https://github.com/openvinotoolkit/mmdetection/pull/260

Ilya-Krylov commented:
@LeonidBeynenson I guess dependency is inversed

Ilya-Krylov commented:
run ote_sdk tests

druzhkov-paul commented on ote_sdk/ote_sdk/algo_backends/test_helpers/__init__.py:
```diff
@@ -0,0 +1 @@
+
```
Are you sure that extra level of nesting (I mean "algo_backends") is necessary? I'd say that "utils" is a better place for this test suite.

LeonidBeynenson replied:
These are test helpers for algo backends tests, not for OTE SDK tests, and I would like to make emphasis on it.
Note that there are many OTE SDK tests, and they have their own test helpers -- see them in the file `ote_sdk/ote_sdk/tests/test_helpers.py`.

Ilya-Krylov replied:
I agree with @druzhkov-paul , `algo_backends` is something unknown for SDK

LeonidBeynenson replied:
IMHO this is very informative.
It is the only place where "algo_backends" are mentioned in OTE SDK, but I do not know a better name for it.

Just "utils" is not sufficient, since these files are not for utils or for help making tests in OTE SDK -- it help making tests for external projects that use OTE SDK as an interface.

LeonidBeynenson replied:
After consideration -- fixed: moved `algo_backends/test_helpers -> test_suite`

druzhkov-paul commented on ote_sdk/ote_sdk/algo_backends/test_helpers/e2e_test_system.py:
```diff
@@ -0,0 +1,102 @@
+# Copyright (C) 2021 Intel Corporation
```
Will the whole test suite work w/o e2e?

LeonidBeynenson replied:
Yes, they should.

LeonidBeynenson replied:
(The e2e wrappers defined in this file are transparent classes/decorators if e2e is not installed)

LeonidBeynenson committed:
Fix isort, flake8, mypy, etc warnings

LeonidBeynenson commented:
run ote_sdk tests

LeonidBeynenson commented:
> @LeonidBeynenson I guess dependency is inversed

Yes, but OTE is a root git repository, so from git point of view this PR depends on mmdetection PR.

LeonidBeynenson committed:
Fix pylint issues

LeonidBeynenson commented:
run ote_sdk tests

LeonidBeynenson commented:
NB: I found out a small bug in my test helpers, will fix it soon.

Ilya-Krylov commented on ote_sdk/ote_sdk/algo_backends/test_helpers/e2e_test_system.py:
```diff
@@ -0,0 +1,108 @@
+# Copyright (C) 2018-2021 Intel Corporation
```
can we use the same license header as we use in ote sdk?

LeonidBeynenson replied:
This is new approved Intel header.
I think we will replace it everywhere, at least it will be in new files.

Ilya-Krylov commented on ote_sdk/ote_sdk/algo_backends/test_helpers/fixtures.py:
```diff
@@ -0,0 +1,342 @@
+MUST be overriden in algo backend's conftest.py file.
+"""
+
+# pylint: disable=redefined-outer-name
```
what did you redefine? Can we avoid using this line in this particular case?

LeonidBeynenson replied:
This is a MUST in the file where fixtures are defined.

Please, note that fixture ideology means that the name of fixture parameter may be the name of another fixture.
From pylint point of view it is "name of parameter redefines the name of function defined above", so it is an error.
To turn off these pylint errors I added this line.

Ilya-Krylov commented on ote_sdk/ote_sdk/algo_backends/test_helpers/e2e_test_system.py:
```diff
@@ -0,0 +1,108 @@
+"""
+The functions in the file generate pytest decorators
+for integrating with e2e test system and the class DataCollector
+that allows to push information to the dashboard in e2e test system.
```
allows pushing / allows us to push

LeonidBeynenson replied:
Fixed

Ilya-Krylov commented on ote_sdk/ote_sdk/algo_backends/test_helpers/fixtures.py:
```diff
@@ -0,0 +1,342 @@
+@pytest.fixture
+def ote_test_domain_fx():
+    """
+    The fixture returns an string that will be used as the 'subject' field in the
```
a string

LeonidBeynenson replied:
Fixed

Ilya-Krylov commented on ote_sdk/ote_sdk/algo_backends/test_helpers/fixtures.py:
```diff
@@ -0,0 +1,342 @@
+    """
+    The fixture returns an string that will be used as the 'subject' field in the
+    e2e test system dashboard.
+    At the moment it is supposed that the fixture should returns something like
```
should return

LeonidBeynenson replied:
Fixed

Ilya-Krylov commented on ote_sdk/ote_sdk/algo_backends/test_helpers/fixtures.py:
```diff
@@ -0,0 +1,342 @@
+@pytest.fixture
+def ote_test_scenario_fx():
+    """
+    The fixture returns an string that will be used as the 'scenario' field in the
```
a string

LeonidBeynenson replied:
Fixed

Ilya-Krylov commented on ote_sdk/ote_sdk/algo_backends/test_helpers/fixtures.py:
```diff
@@ -0,0 +1,342 @@
+    """
+    The fixture returns an string that will be used as the 'scenario' field in the
+    e2e test system dashboard.
+    At the moment it is supposed that the fixture should returns something like
```
should return

LeonidBeynenson replied:
Fixed

Ilya-Krylov commented on ote_sdk/ote_sdk/algo_backends/test_helpers/fixtures.py:
```diff
@@ -0,0 +1,342 @@
+#
+#########################################################################################
+
+# pylint: disable=invalid-name
```
?

LeonidBeynenson replied:
I do not know why pylint supposes that function cannot be uppercase.

Ilya-Krylov replied:
https://www.python.org/dev/peps/pep-0008/#function-and-variable-names

LeonidBeynenson replied:
I really like to use "uppercase with underscore" name for the functions that return constants.  
I think it is obvious and clear.  
Do you think that it should be changed? why? they are just constants indeed.

Ilya-Krylov commented on ote_sdk/ote_sdk/algo_backends/test_helpers/fixtures.py:
```diff
@@ -0,0 +1,342 @@
+
+def ROOT_PATH_KEY():
+    """
+    Constant for storying in dict-s with paths the root path
```
storing

LeonidBeynenson replied:
Fixed

Ilya-Krylov commented on ote_sdk/ote_sdk/algo_backends/test_helpers/fixtures.py:
```diff
@@ -0,0 +1,342 @@
+    {
+        <dataset_name>: {
+            'annotations_train': <annotation_file_path1>
+            'images_train_dir': <images_folder_path1>
```
it limits us to use images only

LeonidBeynenson replied:
Fixed -- added more abstract description

Ilya-Krylov commented on ote_sdk/ote_sdk/algo_backends/test_helpers/fixtures.py:
```diff
@@ -0,0 +1,342 @@
+    Return mapping model names to template paths, received from globbing the
+    folder pointed by the fixture ote_templates_root_dir_fx.
+    Note that the function searches files with name `template.yaml`, and for each such file
+    the model name is the name of the parent folder of the file.
```
let's use model template id https://github.com/openvinotoolkit/training_extensions/blob/develop/ote_sdk/ote_sdk/entities/model_template.py#L325

LeonidBeynenson replied:
Do you think that it is a good idea to fix this and read model_template_id from YAML file?

Ilya-Krylov replied:
I think so, parent folder name must not be meaningful and unique

LeonidBeynenson replied:
Ok, will fix it

LeonidBeynenson replied:
Hmmm, after some reviewing I found out that instead of "gen3_mobilenetV2_ATSS" we will have model name "Custom_Object_Detection_Gen3_ATSS".
Please, note that this model name is much longer.

Also please note that this model name will appear in each test id. At the moment test id-s are sufficiently long, e.g.
`test[ACTION-training_evaluation,model-gen3_mobilenetV2_ATSS,dataset-bbcd,num_iters-KEEP_CONFIG_FIELD_VALUE,batch-KEEP_CONFIG_FIELD_VALUE,usecase-reallife]`

But after this renaming we will have test id-s even longer:
`test[ACTION-training_evaluation,model-Custom_Object_Detection_Gen3_ATSS,dataset-bbcd,num_iters-KEEP_CONFIG_FIELD_VALUE,batch-KEEP_CONFIG_FIELD_VALUE,usecase-reallife]`

@Ilya-Krylov , @druzhkov-paul  -- Is it OK for you??

Ilya-Krylov replied:
not too much difference for me

LeonidBeynenson replied:
Ok.

LeonidBeynenson replied:
Fixed

LeonidBeynenson committed:
Fix issue with wrong default values in training_tests_helper.py

LeonidBeynenson committed:
Fix copyrights and comments

LeonidBeynenson committed:
Add one more pytest option in test helpers

This option is required for the current version of CI only, it should be
removed soon.

LeonidBeynenson committed:
Make OTE SDK test helpers read model name from the field "model_template_id" of model template file

Also improve pylint handling of constant functions.

LeonidBeynenson committed:
Make template*.yaml also be found as template files in OTE SDK training tests

LeonidBeynenson committed:
Merge branch 'develop' into lb/add_ote_sdk_test_helpers

LeonidBeynenson commented:
Resolved conflicts, merged develop to this branch.
Build 171 is run. :runner:

LeonidBeynenson commented:
Build 171 is succeeded ✔️

LeonidBeynenson commented:
2 @druzhkov-paul , @Ilya-Krylov : could you merge this PR as soon as the PR https://github.com/openvinotoolkit/mmdetection/pull/260 is merged?

Ilya-Krylov commented on ote_sdk/ote_sdk/algo_backends/test_helpers/fixtures.py:
```diff
@@ -0,0 +1,361 @@
+#########################################################################################
+
+
+def ROOT_PATH_KEY():  # pylint: disable=invalid-name
```
follow PEP please

LeonidBeynenson replied:
In this case I will have to use usual global variables as constant, whereas I would like to use functions (to avoid changes in the variables, since python does not have `const` modificator).
Do you **really** insist on using variables instead of functions?

LeonidBeynenson replied:
Ok. Fixed

LeonidBeynenson committed:
Move OTE SDK test helpers

Ilya-Krylov commented on .pylintrc:
```diff
@@ -445,7 +445,9 @@ good-names=i,
            ys,
-           i
+           i,
+           p,
```
Can you avoid extending this list?

LeonidBeynenson replied:
I use `f` and `p` as variables in functions.
(`f` -- for opened files, `p` -- for paths)

Ilya-Krylov commented on ote_sdk/ote_sdk/test_suite/fixtures.py:
```diff
@@ -0,0 +1,361 @@
+    data = {}
+    for cur_path in glb:
+        assert osp.isabs(cur_path), f"Error: not absolute path {cur_path}"
+        name = _get_template_name_from_template_file(cur_path)
```
could you use parse_model_template from SDK?

Ilya-Krylov replied:
model_template_id = parse_model_template(cur_path).model_template_id

LeonidBeynenson replied:
Ok

LeonidBeynenson replied:
Fixed

Ilya-Krylov commented on ote_sdk/ote_sdk/test_suite/fixtures.py:
```diff
@@ -0,0 +1,361 @@
+                f"Duplication of names in {root} folder: {data[name]} and {cur_path}"
+            )
+        data[name] = cur_path
+    data[ROOT_PATH_KEY()] = ""
```
what if `name` == `ROOT_PATH_KEY()`

LeonidBeynenson replied:
Ok, will add assert

LeonidBeynenson replied:
Fixed

Ilya-Krylov commented on ote_sdk/ote_sdk/test_suite/fixtures.py:
```diff
@@ -0,0 +1,361 @@
+    return data
+
+
+def _get_template_name_from_template_file(template_path):
```
not need, there is `parse_model_template()`

LeonidBeynenson replied:
Ok

LeonidBeynenson replied:
Fixed

Ilya-Krylov commented on ote_sdk/ote_sdk/test_suite/pytest_insertions.py:
```diff
@@ -0,0 +1,122 @@
+        help="Optional. If the parameter is set, it points the YAML file with expected test metrics.",
+    )
+
+    # TODO(lbeynens): remove it after update CI
```
when?

LeonidBeynenson replied:
When Anton merges his latest changes in CI (in the next week)

Ilya-Krylov commented on ote_sdk/ote_sdk/test_suite/training_test_case.py:
```diff
@@ -0,0 +1,133 @@
+    return dups
+
+
+def _str_dict_with_shortened_vals(d, max_len=200):
```
json.dumps(obj, indent=4) maybe ?

LeonidBeynenson replied:
No, since it could dump a dataset (i.e. all elements of the dataset) -- it will require a lot of lines and it is not required here indeed.

Ilya-Krylov commented on ote_sdk/ote_sdk/test_suite/training_tests_actions.py:
```diff
@@ -0,0 +1,648 @@
+            f"name={self.name}, "
+            f"with_validation={self.with_validation}, "
+            f"depends_stages_names={self.depends_stages_names}"
+            f")"
```
glue to previous line

LeonidBeynenson replied:
Ok

LeonidBeynenson replied:
Fixed

Ilya-Krylov commented on ote_sdk/ote_sdk/test_suite/training_tests_helper.py:
```diff
@@ -0,0 +1,348 @@
+
+            model_dataset_pairs = list(itertools.product(model_names, dataset_names))
+
+            for m, d in model_dataset_pairs:
```
model_name, dataset_name

LeonidBeynenson replied:
Ok

LeonidBeynenson replied:
Fixed

Ilya-Krylov commented on ote_sdk/ote_sdk/test_suite/training_tests_common.py:
```diff
@@ -0,0 +1,75 @@
+    return score.name, score.value
+
+
+def convert_hyperparams_to_dict(hyperparams):
```
what is wrong with https://github.com/openvinotoolkit/training_extensions/blob/develop/ote_sdk/ote_sdk/configuration/helper/convert.py#L105 ?

LeonidBeynenson replied:
Remove it, it is not used anymore

LeonidBeynenson replied:
Fixed

Ilya-Krylov commented on ote_sdk/ote_sdk/test_suite/training_tests_common.py:
```diff
@@ -0,0 +1,75 @@
+# pylint: disable=invalid-name
+
+
+def DEFAULT_FIELD_VALUE_FOR_USING_IN_TEST():
```
please follow PEP convention or just use CONSTANTS

LeonidBeynenson replied:
OMG, Ilya, do you **really** think that it is a good idea to use variables instead of functions that return constant values?

LeonidBeynenson replied:
Fixed

Ilya-Krylov commented on ote_sdk/ote_sdk/test_suite/training_tests_common.py:
```diff
@@ -0,0 +1,75 @@
+    return "DEFAULT_FIELD_VALUE_FOR_USING_IN_TEST"
+
+
+def KEEP_CONFIG_FIELD_VALUE():
```
ditto

LeonidBeynenson replied:
Fixed

Ilya-Krylov commented on ote_sdk/ote_sdk/test_suite/training_tests_common.py:
```diff
@@ -0,0 +1,75 @@
+    return "KEEP_CONFIG_FIELD_VALUE"
+
+
+def REALLIFE_USECASE_CONSTANT():
```
ditto

LeonidBeynenson replied:
Fixed

Ilya-Krylov commented on ote_sdk/ote_sdk/test_suite/training_tests_helper.py:
```diff
@@ -0,0 +1,348 @@
+        self.short_test_parameters_names_for_generating_id = (
+            test_creation_parameters.short_test_parameters_names_for_generating_id()
+        )
+        # TODO(beynens): rename to test_parameters_defining_test_case
```
?

LeonidBeynenson replied:
Remove

LeonidBeynenson replied:
Fixed

Ilya-Krylov commented on ote_sdk/ote_sdk/test_suite/fixtures.py:
```diff
@@ -0,0 +1,361 @@
+    ), f"Error: ote_templates_root_dir_fx is not an absolute path: {root}"
+    glb = glob.glob(f"{root}/**/template*.yaml", recursive=True)
+    data = {}
+    for cur_path in glb:
```
please use something meaningful instead of `glb`

LeonidBeynenson replied:
Fixed

Ilya-Krylov commented on ote_sdk/ote_sdk/test_suite/fixtures.py:
```diff
@@ -0,0 +1,361 @@
+#########################################################################################
+
+
+def ROOT_PATH_KEY():  # pylint: disable=invalid-name
```
follow PEP conventions or use CONSTANT

LeonidBeynenson replied:
Fixed

Ilya-Krylov commented:
as soon as comments are resolved, 'run ote_sdk tests' must be written as a comment and the branch must be checked on BMM

LeonidBeynenson committed:
Make decorative changes w.r.t. PR comments

LeonidBeynenson commented:
run ote_sdk tests

LeonidBeynenson commented:
Build 190 is run 🏃 
(It was succeeded ✔️ )

LeonidBeynenson committed:
Merge branch 'develop' into lb/add_ote_sdk_test_helpers

LeonidBeynenson commented:
Resolved conflicts, merge develop branch to this branch.
The build 193 is run. 🏃

LeonidBeynenson commented:
2 @Ilya-Krylov , @druzhkov-paul , please let's merge this PR if the build 193 succeeds.

(And note that it should be done after PR https://github.com/openvinotoolkit/mmdetection/pull/260 is merged: when that PR is merged, this PR should be updated w.r.t. the merge, and then this PR should be merged ASAP, otherwise CI will not work.)

Tomorrow I'll be ready for the merges and I really ask you not to merge other PR-s before, since otherwise the next round of the "Achilles and the Tortoise" game will be started.

LeonidBeynenson commented:
run ote_sdk tests

Ilya-Krylov commented:
run ote_sdk tests

Ilya-Krylov commented:
@LeonidBeynenson 
Order should be different:
1. Merge this PR
2. Merge mmdetection PR
3. New PR to OTE with mmdetection update pointing to `dp/nous_sdk_api` head (actually it is optional but desirable, as soon as 2. is merged without changing commit hash)

Ilya-Krylov commented:
please update your mmdet PR.

Ilya-Krylov commented:
after that this PR must be tested on BMM

LeonidBeynenson committed:
Update external/mmdetection submodule

LeonidBeynenson committed:
Merge branch 'develop' into lb/add_ote_sdk_test_helpers

LeonidBeynenson commented:
Build 193 passed ✔️ 
Merged the develop branch to this branch
Build 195 is run 🏃

LeonidBeynenson commented:
run ote_sdk tests

LeonidBeynenson commented:
Build 195 is failed in `end_to_end` test due to random nature of the test.
Build 196 is run. 🏃

Ilya-Krylov commented:
@LeonidBeynenson let's increase the threshold
```
 E   AssertionError: 0.0022222222222223476 not less than or equal to 0.0005 : Too big performance difference after OpenVINO export. (reference metric: 1.0, actual value: 0.9977777777777777, delta tolerance threshold: 0.0005)
```
If you think that it is not proper way - one should report this problem to openvino team.

Ilya-Krylov commented:
BMM 21123 Tests passed
